prow/gerrit/gerrit.go

```diff
@@ -0,0 +1,210 @@
+		if err := c.ProcessChange(change); err != nil {
+			logrus.WithError(err).Errorf("Failed process change %v", change.CurrentRevision)
+		}
+	}
+
+	c.lastUpdate = syncTime
+	if err := c.SaveLastSync(syncTime); err != nil {
+		logrus.WithError(err).Errorf("last sync %v, cannot save to path %v", syncTime, c.storage)
+	}
+	logrus.Infof("Processed %d changes", len(changes))
+	return nil
+}
+
+// QueryChanges will query all gerrit changes since controller's last sync loop
+func (c *Controller) QueryChanges() (map[string]gerrit.ChangeInfo, error) {
+	opt := &gerrit.QueryChangeOptions{}
+	for _, proj := range c.projects {
+		opt.Query = append(opt.Query, "project:"+proj)
+	}
+	opt.AdditionalFields = []string{"CURRENT_REVISION"}
+
+	// store a map of changeID:change
+	pending := map[string]gerrit.ChangeInfo{}
+	start := 0
+
+	for {
+		opt.Limit = c.ca.Config().Gerrit.RateLimit
+		opt.Start = start
+
+		// The change output is sorted by the last update time, most recently updated to oldest updated.
+		// Gerrit API docs: https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes
+		changes, _, err := c.gc.QueryChanges(opt)
+		if err != nil {
+			// should not happen? Let next sync loop catch up
+			return pending, fmt.Errorf("failed to query gerrit changes: %v", err)
+		}
+
+		if len(*changes) == 0 {
+			return pending, nil
+		}
+		start += len(*changes)
+
+		for _, change := range *changes {
+			// if we already processed this change, then we stop the current sync loop
+			layout := "2006-01-02 15:04:05"
+			updated, err := time.Parse(layout, change.Updated)
+			if err != nil {
+				logrus.WithError(err).Error("Parse time %v failed", change.Updated)
+				continue
+			}
+
+			// process if updated later than last updated
+			// stop if already parsed
+			if updated.After(c.lastUpdate) {
+				// in case of a new change comes in during the current sync cycle
+				if _, ok := pending[change.CurrentRevision]; !ok {
+					pending[change.CurrentRevision] = change
+				}
+			} else {
+				return pending, nil
+			}
+		}
+	}
+}
+
+// ProcessChange creates new presubmit prowjobs base off the gerrit changes
+func (c *Controller) ProcessChange(change gerrit.ChangeInfo) error {
+	rev, ok := change.Revisions[change.CurrentRevision]
+	if !ok {
+		return fmt.Errorf("cannot find current revision for change %v", change.ID)
+	}
+
+	logger := logrus.WithField("gerrit change", change.Number)
+
+	for _, spec := range c.ca.Config().Presubmits[change.Project] {
+		kr := kube.Refs{
+			Org:     c.instance,
+			Repo:    change.Project,
+			BaseRef: "",
+			BaseSHA: change.BaseChange,
+			Pulls: []kube.Pull{
+				{
+					Number: change.Number,
+					Author: change.Owner.Name,
+					SHA:    rev.Ref,
+				},
+			},
+		}
+		pj := pjutil.NewProwJob(pjutil.PresubmitSpec(spec, kr), map[string]string{})
+		logger.WithFields(pjutil.ProwJobFields(&pj)).Info("Creating a new prowjob for.")
+		if _, err := c.kc.CreateProwJob(pj); err != nil {
```

Comment:
Since we trigger PJs without regard for RunIfChanged, RunAfterSuccess, or AlwaysRun we should enforce that these fields are not used with a unit test.

Reply:
IMO Plank will handle RunAfterSuccess right? I'll later on add support for `AlwaysRun` and `RunIfChanged`

Reply:
added a TODO for AlwaysRun and RunIfChanged. In Gerrit I think everything will be AlwaysRun for now, as there's no plugin support to manually trigger tests yet.